specs/api_cron_flow.md

```diff
@@ -0,0 +1,130 @@
+from the SBA.
+
+The first is an API to get the Yelp rating where possible.
+
+The second is a Google Civics API to get the Congressional district
+based on the borrower address.
+
+The third is an attempt to obtain the latitude and longitude for the
+borrower addresses.
+
+Given the size of our database (even just for the SFDO loan set), if
+we try to update all the ratings in a single run, the API providers
+may throttle our access. The Google Civics API has a daily limit of
+25,000 calls and one of the potential goecode APIs has a 2,500 daily
+limit.
+
+The request is to be able to chunk the output into small batches and
+schedule a cron job to run repeatedly until all records have been
+attempted.
+
+## Final Result
+
+For this project, we will make changes to the script to allow it to
+run a specific set of records and then exit. We will provide the
+capability to keep track of which records have been attempted and
+which have not. We will provide the ability to update the external
+APIs separately based on different limitations. We will also provide
+an example of a cron scheduling job that would permit the entire
+database to be refreshed over a period of hours or days.
+
+## Proposed Approach
+
+### General Discussion of Approach
+
+We discussed an approach whereby a wrapper script is used to setup the
+database and Python environment. The script will then pass arguments
+down to the control script, which will select records to be processed
+by each API. The control script will pass a selection of sba_sfdo_id
+fields (or other appropriate key) to the execution script for each API
+that is to be run.
+
+This approach means that the API execution script does not need to
+worry about processing the arguments or filtering the records.  Other
+than updating the timestamp, the execution code has very little
+knowledge of the batch control process.
+
+The cron job will call the wrapper script with the appropriate
+args. There will be a number of entries in the cron file to properly
+process all the external APIs in a timely manner.
+
+### Updates to the API control script
+
+The API control script will support the following new arguments.
+
+At least one of the following must be specified (all three may be
+specified if desired):
+
+--yelp: indicates that the script should update the Yelp entries.
+
+--civics: indicates that the script should update the Congressional
+  district entries.
+
+--geocode: indicates tha the script should update the Geocode
+  entries. Optional, if not specified these records are unchanged.
+
+The following are optional but recomended:
+
+--max_attempts <integer>: indicates the maximum number of records to
+  select for updating. The program will select records which have
+  never been updated first, and then sort others by last update
+  time. If not specified, all records will be selected for update,
+  subject to the --older_than value.
+
+--older_than <days>: indicates that only records older than the
```

Comment:
This is great.

Reply:
Should allow us to recycle the records to avoid storing beyond the usage license.